Pool objects caught by the catcher

ekrctb opened the pull request:
Addressing the second homework mentioned in my previous PR #11011.

Before this PR, when a hit object is caught by the catcher, a DHO was created and placed on the catcher plate. Caught objects are constantly created and destroyed, and the allocation overhead is significant. Pooling caught objects is desired to reduce stuttering during the gameplay.

However, several DHO logics that are unneeded for a static caught object interfere with the desired pooling logic. For example, we had to use `ApplyCustomState` to workaround an issue of transforms cleared by DHO.

Thus, I decided to create a new class `CaughtObject` for a caught object, not a derived class of DHO.
To reduce code duplication, I extracted common properties of `DrawablePalpableCatchHitObject` and `CaughtObject` to `IHasCatchObjectState` interface.

Visual properties such as `AccentColour` are copied from the DHO to `CaughtObject`, not bound.
A known issue is if the skin is changed then caught objects will keep old skin's combo colors.
 The issue is properties such as `AccentColour` are only present in DHO, not in HOs. I'm currently not sure how to solve this issue. We cannot bind to a DHO because the DHO will be reused for another hit object while the `CaughtObject` is still alive.

I refactored some of the object catching logic of the catcher.
<s>I removed `ScaleContainer` from catch DHO to make the implementation simpler.</s>
Edit: I added it back.

ekrctb committed:
Compute object position in stack via a pure function

ekrctb committed:
Remove ScaleContainer and flatten the Drawable tree of catch DHO

ekrctb committed:
Make object on the catcher plate separate CaughtObject class

ekrctb committed:
Make catch hit lighting logic not dependent on caught object

ekrctb committed:
Fix hit lighting colour not applied in TestSceneCatcher

ekrctb committed:
Introduce IHasCatchObjectState implemented by DHO and CaughtObject

ekrctb committed:
Pool caught objects and dropped objects

ekrctb committed:
Fix dropped objects not removed on revert result

ekrctb committed:
Add doc comments a bit

ekrctb committed:
Rename to CopyStateFrom, and add comment

peppy commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObject.cs:
```diff
@@ -0,0 +1,100 @@
+        }
+    }
+
+    public class CaughtFruit : CaughtObject, IHasFruitState
```
Probably should move these to their own files (even for small classes we generally do this as a rule).

bdach commented on osu.Game.Rulesets.Catch/UI/Catcher.cs:
```diff
@@ -441,138 +458,123 @@ private void updateState(CatcherAnimationState state)
         {
-            if (!caughtFruitContainer.Remove(caughtObject))
-                throw new InvalidOperationException("Can only drop a caught object on the plate");
+            var caughtObjects = caughtFruitContainer.Children.ToArray();
+            var droppedObjects = caughtObjects.Select(getDroppedObject).ToArray();
```
Not sure about this double copy. Is there a reason to not just immediately `.Select(getDroppedObject)`, therefore allocating just one array?

bdach commented on osu.Game.Rulesets.Catch/UI/Catcher.cs:
```diff
@@ -441,138 +458,123 @@ private void updateState(CatcherAnimationState state)
-            d.Anchor = Anchor.TopLeft;
-            d.Position = caughtFruitContainer.ToSpaceOfOtherDrawable(d.DrawPosition, droppedObjectTarget);
+            if (!caughtFruitContainer.Remove(caughtObject))
+                throw new InvalidOperationException("Can only drop a caught object on the plate");
```
Not sure if I'm understanding the intent right, would maybe reword to
```suggestion
                throw new InvalidOperationException("Can only drop objects that were previously caught on the plate");
```

ekrctb replied:
Actually this check is not needed anymore because this method was public but it is now private and the exception shouldn't occur.

ekrctb commented on osu.Game.Rulesets.Catch.Tests/TestSceneCatcherArea.cs:
```diff
@@ -73,7 +73,10 @@ private void createCatcher(float size)
             SetContents(() =>
             {
-                var droppedObjectContainer = new Container();
+                var droppedObjectContainer = new Container
+                {
+                    RelativeSizeAxes = Axes.Both
+                };
```
I had a bug of not setting `Anchor` of `CaughtObject` when the object is dropped. The bug was not apparent in this visual test when the container is size 0 (not setting `RelativeSizeAxes`) so I added it for the sake of robustness of the visual test.

ekrctb commented on osu.Game.Rulesets.Catch/Objects/Drawables/DrawablePalpableHasCatchHitObject.cs:
```diff
@@ -5,60 +5,49 @@
 {
-    public abstract class DrawablePalpableCatchHitObject : DrawableCatchHitObject
+    [Cached(typeof(IHasCatchObjectState))]
+    public abstract class DrawablePalpableHasCatchHitObject : DrawableCatchHitObject, IHasCatchObjectState
```
This is an error from Rider renaming feature. I'll fix it.

ekrctb commented on osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObject.cs:
```diff
@@ -0,0 +1,100 @@
+            ClearTransforms();
+
+            Alpha = 1;
+            LifetimeStart = double.MinValue;
+            LifetimeEnd = double.MaxValue;
```
Lifetime is set by `Expire` when a caught object is dropped. If lifetime is not reset issues will occur.
I removed assignment of `LifetimeStart` because that was not doing anything (I thought the object will be removed when a replay is rewound but it is not, and I added object removing logic to `OnRevertResult`)

ekrctb replied:
I somehow thought lifetime had to reset but apparently it is reset by `DrawablePool` and the lines above is not necessary. I think I misunderstood a different bug (that I solved).

ekrctb commented on osu.Game.Rulesets.Catch/Skinning/Default/CatchHitObjectPiece.cs:
```diff
@@ -54,8 +47,8 @@ protected override void LoadComplete()
         {
-            if (BorderPiece != null && DrawableHitObject?.HitObject != null)
-                BorderPiece.Alpha = (float)Math.Clamp((DrawableHitObject.HitObject.StartTime - Time.Current) / 500, 0, 1);
+            if (BorderPiece != null && ObjectState?.HitObject != null)
```
For a DHO, `HitObject` can be null here for some reason.

ekrctb committed:
Revert rename error

ekrctb committed:
Remove null check that is not required anymore

ekrctb committed:
more specific type droppedObjectContainer

ekrctb committed:
Rename caughtFruitContainer -> caughtObjectContainer

ekrctb committed:
Remove unnecessary copy

ekrctb committed:
Reword exception message

ekrctb committed:
Move classes to separate files

ekrctb committed:
Remove unneeded lifetime assignment

ekrctb committed:
Remove unneeded check of caught object removal

The logic was public but now it is private
so the condition is ensured by the caller

bdach reviewed (dismissed):
Looks good to go as far as I can tell :+1:

bdach commented:
I think the accent colour issue can be resolved in a separate pull. An option would be to have `AccentColour` in the hitobjects themselves and associate `CaughtObject`s with them, but I imagine managing that state could turn ugly...

peppy commented on osu.Game.Rulesets.Catch/Objects/Drawables/DrawableBanana.cs:
```diff
@@ -44,12 +44,12 @@ protected override void UpdateInitialTransforms()
 
-            ScaleContainer.ScaleTo(HitObject.Scale * (end_scale + random_scale_range * RandomSingle(3)))
-                          .Then().ScaleTo(HitObject.Scale * end_scale, HitObject.TimePreempt);
+            this.ScaleTo(HitObject.Scale * (end_scale + random_scale_range * RandomSingle(3)))
```
This is probably not what we want to be doing. The rationale behind not adjusting one's own scale is that mods may want to do so as well, causing potentially loss (or otherwise incorrect) transforms.

Can you add back the local scale container and instead forward the interface implementation to the `HitObject` model's scale (can be done implicitly to avoid conflict with the underlying `Drawable.Scale`)?

peppy commented on osu.Game.Rulesets.Catch/Objects/Drawables/DrawableDroplet.cs:
```diff
@@ -26,9 +24,9 @@ public DrawableDroplet([CanBeNull] CatchHitObject h)
         private void load()
         {
-            ScaleContainer.Child = new SkinnableDrawable(
+            AddInternal(new SkinnableDrawable(
```
As above.

ekrctb committed:
Add ScalingContainer back

Don't want to set DHO.Scale or DHO.Rotation because because DHO may be transformed by mods.
DHO.Size is also assigned for drawable visualizer

ekrctb committed:
Combine DisplayRadius and Scale to DisplaySize

ekrctb commented:
My commit above changed the radius of banana considered for the caught object stacking. It is now more correct and matches visual (previously scaling transform was ignored for the radius).

peppy committed:
Add some spacing to interface class

peppy commented on osu.Game.Rulesets.Catch/UI/Catcher.cs:
```diff
@@ -124,7 +134,11 @@ public Catcher([NotNull] Container trailsTarget, [NotNull] Container droppedObje
             {
                 hitExplosionPool = new DrawablePool<HitExplosion>(10),
-                caughtFruitContainer = new Container<DrawablePalpableCatchHitObject>
+                caughtFruitPool = new DrawablePool<CaughtFruit>(50),
```
As a safety to avoid asynchronous loading, we'd usually put these inside the `[BackgroundDependencyLoader]` method, but I've checked this case is fine here (async happens from `CreateDrawableRuleset` and includes all `ctor`s).

peppy reviewed (approved):
Looks good here!
